### ISSUE TITLE
Index out of bound for Urban LCZ cases in conversion of lfmass

### ISSUE THREAD

**Issue by bfkg** [NCAR/noahmp#104]
If urban with LCZ is selected IVGTYP ranges from 1-61. SLA_TABLE is however only defined up to 20/27 depending on the landuse dataset selected. This leads to reading beyond the bounds of SLA_TABLE.

```fortran
             masslai = 1000. / max(SLA_TABLE(IVGTYP(I,J)),1.0)
```

https://github.com/NCAR/noahmp/blob/981d4f859ce6c64213d38a783654c05b47b3485e/drivers/wrf/module_sf_noahmpdrv.F#L2184C14-L2184C101

Maybe something like this can solve the issue:

``` fortran
if(urbanpt_flag)  then
   masslai = 1000. / max(SLA_TABLE(NATURAL_TABLE),1.0)
else 
   masslai = 1000. / max(SLA_TABLE(IVGTYP(I,J)),1.0)
endif

**Comment by bfkg** [comment]
Will the WRF driver also be updated for the next WRF release?

**Comment by cenlinhe** [comment]
I am not sure why WRF seems doing OK with this issue even without this fix.
@tslin2 Have you run into any problems using WRF-urban-Noah-MP with LCZ for this issue: https://github.com/NCAR/noahmp/blob/981d4f859ce6c64213d38a783654c05b47b3485e/drivers/wrf/module_sf_noahmpdrv.F#L2184

It seems that we also need to fix it in the non-refactored WRF/NoahMP code using the same above fix that has been included in the refactor code.

**Comment by tslin2** [comment]
Hi, @cenlinhe and @bfkg 

I don't encounter any issue in WRF-Noah-MP-urban with LCZ.
I made a quick test.
SLA_TABLE(IVGTYP(I,J)) was zero , so masslai was 1000.0 using LCZ

**Comment by tslin2** [comment]
Hi @cenlinhe,


https://github.com/NCAR/noahmp/blob/6c9717b6710f30ff7217ebc51cd93e2fa578d516/drivers/hrldas/NoahmpInitMainMod.F90#L190-L196


Look like this fix should use urbanpt_flag as @bfkg suggested for the if statement 
because there could be not urban types such as in rural grids with WRF-urban model turn on (NoahmpIO%SF_URBAN_PHYSICS /= 0).

I am also thinking the NATURAL_TABLE should change to ISURBAN_TABLE, so the result can be the same as the original single urban type when LCZ is not used but WRF-urban model turn on.


if(urbanpt_flag)  then
   masslai = 1000. / max(SLA_TABLE(ISURBAN_TABLE),1.0)
else 
   masslai = 1000. / max(SLA_TABLE(IVGTYP(I,J)),1.0)
endif

**Comment by cenlinhe** [comment]
@tslin2 I agree with you that we should use urbanpt_flag, but I do not think we should use "ISURBAN_TABLE" because the idea here is that if urban physics is turned on and Noah-MP column model here will simulate the rural portion of the urban grid. We should not use the bulk urban type "ISURBAN_TABLE" for the SLA parameter.

**Comment by cenlinhe** [comment]
I will try to fix this issue in both HRLDAS and WRF versions this week or next week.

**Comment by tslin2** [comment]
got it, thanks!

**Comment by cenlinhe** [comment]
I have fixed this bug in Noah-MP v4.5 at this commit: https://github.com/NCAR/noahmp/commit/6f2bf7c575a0099d60b51efe8811c6519d531abe
I have also submitted a WRF PR to include this fix: https://github.com/wrf-model/WRF/pull/1971

**Comment by cenlinhe** [comment]
I am going to close this issue since it has been solved.

**Comment by tslin2** [comment]
should this also be fixed in the version 5?!

**Comment by cenlinhe** [comment]
I will fix this in version 5.0 very soon too.
